content/4_Gold/Sliding_Window.mdx

```diff
@@ -295,7 +295,189 @@ def maxSlidingWindow(self, nums: List[int], k: int) -> list[int]:
+				s2.push({s1.top().first, max(mx, s1.top().first)});
+				s1.pop();
+			}
+		}
+		s2.pop();
+	}
+};
+
+class Solution {
+public:
+	vector<int> maxSlidingWindow(vector<int> &nums, int k) {
+		MaxQueue q;
+		vector<int> max_vals;
+
+		// Fill the queue with elements from the first window
+		for (int i = 0; i < k; i++) {
+			q.enqueue(nums[i]);
+		}
+		max_vals.push_back(q.query());
+
+		/*
+		 * We slide the window to the right by removing the first element
+		 * from the queue and adding the new element at the end of the queue.
+		 * For each window, we add the maximum to our result array max_vals.
+		 */
+		for (int i = k; i < nums.size(); i++) {
+			q.dequeue();
+			q.enqueue(nums[i]);
+			max_vals.push_back(q.query());
+		}
+
+		return max_vals;
+	}
+};
+```
+</CPPSection>
+
+
+<JavaSection>
+```java
+class Solution {
+	public int[] maxSlidingWindow(int[] nums, int k) {
+		MaxQueue q = new MaxQueue();
+		int n = nums.length;
+		int[] maxVals = new int[n - k + 1];
+
+		// Fill the queue with elements from the first window
+		for (int i = 0; i < k; i++) {
+			q.enqueue(nums[i]);
+		}
+		maxVals[0] = q.query();
+
+		/*
+		 * We slide the window to the right by removing the first element
+		 * from the queue and adding the new element at the end of the queue.
+		 * For each window, we add the maximum to our result array maxVals.
+		 */
+		for (int i = k; i < n; i++) {
+			q.dequeue();
+			q.enqueue(nums[i]);
+			maxVals[i - k + 1] = q.query();
+		}
+
+		return maxVals;
+	}
+
+	private static class MaxQueue {
+		/*
+		 * For each Pair e, e.a is the value of the element and
+		 * e.b is the maximum among all elements in that stack under element e.
+		 */
+		public Stack<Pair> s1, s2;
+
+		public MaxQueue() {
+			s1 = new Stack<>();
+			s2 = new Stack<>();
+		}
+
+		/*
+		 * Get the maximum element in the MaxQueue.
+		 * It is the maximum of both stacks which are stored in s.peek().b.
+		 */
+		public int query() {
+			if (s1.empty() && s2.empty()) {
+				return -1000000000;
+			}
+			if (s1.empty() || s2.empty()) {
+				return s1.empty() ? s2.peek().b : s1.peek().b;
+			}
+			return Math.max(s1.peek().b, s2.peek().b);
+		}
+
+		/*
+		 * Add a new element into our MaxQueue. We add the value of this
+		 * element itself and the maximum element in the stack s1 after adding.
+		 */
+		public void enqueue(int val) {
+			s1.push(new Pair(val, s1.empty() ? val : Math.max(val, s1.peek().b)));
+		}
+
+		/*
+		 * Remove the first element from our MaxQueue by poping the top element from
+		 * s2. Since elements in s1 are stored in the order of how we added them, i.e.
+		 * the last added element is at the top of s1, we just have to pop all of them
+		 * out and push them into the stack s2. After that, these elements in s2 will
+		 * be in reversed order, i.e. the first added element will be at the top of s2
+		 * and we can pop them out as from normal stacks. We only move elements from
+		 * s1 to s2 when s2 is empty.
+		 */
```

Comment:
at this point just put it above in an explanation, comments in guide code shouldn't be this long